File: LAB QUIZZES/CSE221_Quiz02/_A2A/_A2A.java
import java.util.*;
import java.io.*;

public class _A2A {
    public static Random rng = new Random();
    public static int test = 0, best = 0, score = 0, batch = 0, total = 0;
    public static int timeLimit = 1, nBatch = 10;
    public static int weight[] = { 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1 };
    public static int nTest[] = { 0, 10, 100, 100, 100, 100, 100, 100, 10, 10, 10 };
    public static int maxN[] = { 0, 10, 10, 100, 100, 1000, 1000, 10000, 100000, 1000000, 1000000 };
    public static int maxS[] = { 0, 1_000, 1_000_000, 1_000_000, 1_000_000_000, 1_000_000, 1_000_000_000, 1_000_000,
            1_000_000_000, 1_000_000, 1_000_000_000 };
    public static ArrayList<Integer> A = new ArrayList<>();
    public static Integer S = 0;
    public static Integer N = 0;

    public static _A2A_Solution.Pair<Integer, Integer> res;

    public static void printCase(String verdict, int test, Integer N, Integer S, ArrayList<Integer> A)
            throws Exception {
        if (new File(verdict + ".txt").length() == 0) {
            BufferedWriter fout = new BufferedWriter(new FileWriter(verdict + ".txt"));
            fout.write("batch = " + batch + "\ntest = " + test + "\n");
            fout.write("N = " + N + "\n");
            fout.write("S = " + S + "\n");
            fout.write("A = " + A + "\n");
            fout.close();
        }
        System.out.print(verdict + " on Batch " + batch + "\n");
    }

    public static void exitCase(String verdict) throws Exception {
        printCase(verdict, test, N, S, A);
        System.exit(1);
    }

    public static boolean validIndex(int i) {
        return i > 0 && i <= N;
    }

    public static void checkSoln() throws Exception {
        if (res.key != -1) {
            if (!validIndex(res.key) || !validIndex(res.value) || res.key == res.value)
                exitCase("WrongAnswer");
            if (A.get(res.key - 1) + A.get(res.value - 1) != S)
                exitCase("WrongAnswer");
        } else {
            for (int i = 0, j = N - 1; i < j; i += (A.get(i) + A.get(j) > S ? (--j > 0 ? 0 : 0) : 1))
                if (A.get(i) + A.get(j) == S)
                    exitCase("WrongAnswer");
        }
    }

    public static void processBatch() throws Exception {
        for (test = 1; test <= nTest[batch]; ++test) {
            N = 2 + rng.nextInt(maxN[batch] - 1);
            A.clear();
            int D = maxS[batch] / N;
            for (int i = 0; i < N; ++i)
                A.add((i == 0 ? 1 + rng.nextInt(maxS[batch] - (N - i - 1) * D)
                        : A.get(i - 1) + rng.nextInt(maxS[batch] - (N - i - 1) * D - A.get(i - 1) + 1)));
            boolean OK = (rng.nextInt(2) == 1);
            if (OK) {
                int i = rng.nextInt(N);
                int j = i;
                while (i == j)
                    j = rng.nextInt(N);
                S = A.get(i) + A.get(j);
            } else {
                S = 1 + rng.nextInt(maxS[batch]);
            }
            res = _A2A_Solution.solve(N, S, new ArrayList(A));
            checkSoln();
        }
    }

    public static void limitTime() {
        try {
            Thread.sleep(1000 * timeLimit);
            exitCase("TimeLimitExceeded");
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    public static void main(String args[]) throws Exception {
        if (args.length == 1)
            rng.setSeed(batch = Integer.parseInt(args[0]));
        if (1 <= batch && batch <= nBatch) {
            System.out.println("Running on Batch " + batch);
            long start = System.nanoTime();
            new Thread(() -> limitTime()).start();
            processBatch();
            long finish = System.nanoTime();
            long elapsed = finish - start;
            System.out.printf("Passed Batch " + batch + " in %.9fs\n", elapsed * 1e-9);
            System.exit(0);
        }
        BufferedReader fin = new BufferedReader(new FileReader("EnterIDandLanguage.txt"));
        String ID = fin.readLine().split("\\s+", 2)[0];
        fin.close();
        if (new File("_A2A_" + ID + ".java").exists()) {
            fin = new BufferedReader(new FileReader("_A2A_" + ID + ".java"));
            best = Integer.parseInt(fin.readLine().split("\\s+", 4)[2]);
            fin.close();
        }
        new FileOutputStream("WrongAnswer.txt").close();
        new FileOutputStream("TimeLimitExceeded.txt").close();
        for (batch = 1; batch <= nBatch; total += weight[batch], ++batch) {
            Process process = new ProcessBuilder("java", "_A2A", "" + batch).start();
            fin = new BufferedReader(new InputStreamReader(process.getInputStream()));
            fin.lines().forEach(System.out::println);
            fin.close();
            fin = new BufferedReader(new InputStreamReader(process.getErrorStream()));
            fin.lines().forEach(System.out::println);
            fin.close();
            if (process.waitFor() == 0)
                score += weight[batch];
        }
        if (best <= score) {
            BufferedWriter fout = new BufferedWriter(new FileWriter("_A2A_" + ID + ".java"));
            fout.write("// " + ID + " " + score + "\n");
            fout.close();
            new ProcessBuilder("cmd", "/c", "echo // %COMPUTERNAME% %USERNAME%>>_A2A_" + ID + ".java").start()
                    .waitFor();
            fout = new BufferedWriter(new FileWriter("_A2A_" + ID + ".java", true));
            fin = new BufferedReader(new FileReader("_A2A_Solution.java"));
            fout.write(fin.lines().collect(java.util.stream.Collectors.joining(System.lineSeparator())));
            fin.close();
            fout.close();
        }
        System.out.print("Tentative score = " + (score / (double) total) + "/1\n");
        System.exit(0);
    }
}

File: LAB QUIZZES/CSE221_Quiz02/_A2A/_A2A.py
import os, sys, time, random, threading, subprocess
from _A2A_Solution import solve
test, best, score, batch, total = 0, 0, 0, 0, 0
timeLimit, nBatch = 1, 10
weight = [0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]
nTest = [0, 10, 100, 100, 100, 100, 100, 100, 10, 10, 10]
maxN = [ 0, 10, 10, 100, 100, 1000, 1000, 10000, 100000, 1000000, 1000000]
maxS = [0, 1_000, 1_000_000, 1_000_000, 1_000_000_000, 1_000_000, 1_000_000_000, 1_000_000, 1_000_000_000, 1_000_000, 1_000_000_000]
N, S = 0, 0
A = []
res = None
def printCase(verdict, test, N, S, A):
    if os.path.getsize(verdict + ".txt") == 0:
        with open(verdict + ".txt", "w") as fout:
            fout.write("batch = " + str(batch) + "\ntest = " + str(test) + "\n")
            fout.write("N = " + str(N) + "\n")
            fout.write("S = " + str(S) + "\n")
            fout.write("A = " + str(A) + "\n")
    print(verdict + " on Batch " + str(batch), flush=True)
def exitCase(verdict):
    printCase(verdict, test, N, S, A)
    os._exit(1)
def validIndex(i):
    return i > 0 and i <= N
def checkSoln():
    if(len(res) != 2):
        exitCase("WrongAnswer")
    if(res[0] != -1):
        if(not validIndex(res[0]) or not validIndex(res[1]) or res[0] == res[1]):
            exitCase("WrongAnswer")
        if(A[res[0] - 1] + A[res[1] - 1] != S):
            exitCase("WrongAnswer");
    else:
        i, j = 0, N - 1
        while i < j: 
            if(A[i] + A[j] == S):
                exitCase("WrongAnswer")
            if A[i] + A[j] > S:
                j -= 1
            else:
                i += 1
def processBatch():
    global test, N, S, A, res
    test = 1
    while test <= nTest[batch]:
        N = random.randint(2, maxN[batch])
        A = [None] * N
        D = maxS[batch] // N
        for i in range(N):
            A[i] = (random.randint(1 if i == 0 else A[i - 1], maxS[batch] - (N - i - 1) * D))
        OK = (random.randint(0, 1) == 1)
        if(OK):
            i = random.randint(0, N - 1)
            j = i
            while(i == j):
                j = random.randint(0, N - 1)
            S = A[i] + A[j]
        else:
            S = random.randint(1, maxS[batch])
        res = solve(N, S, A.copy())
        checkSoln()
        test += 1
        
def limitTime():
    time.sleep(timeLimit)
    exitCase("TimeLimitExceeded")
if len(sys.argv) == 2:
    batch = int(sys.argv[1])
    random.seed(batch)
if 1 <= batch and batch <= nBatch:
    print("Running on Batch " + str(batch), flush=True)
    start = time.time()
    threading.Thread(target=limitTime, daemon=True).start()
    processBatch()
    finish = time.time()
    elapsed = finish - start
    print("Passed Batch " + str(batch) + " in " + f"{elapsed:.9f}" + "s", flush=True)
    os._exit(0)
ID = ""
with open("EnterIDandLanguage.txt") as fin:
    ID = fin.readline().split()[0]
if os.path.exists("_A2A_" + ID + ".py"):
    with open("_A2A_" + ID + ".py") as fin:
        best = int(fin.readline().split()[2])
open("WrongAnswer.txt", "w").close()
open("TimeLimitExceeded.txt", "w").close()
batch = 1
while batch <= nBatch:
    process = subprocess.run("pypy _A2A.py " + str(batch), capture_output=True, text=True)
    if process.stdout:
        print(process.stdout, end="")
    if process.stderr:
        print(process.stderr, end="")
    if process.returncode == 0:
        score += weight[batch]
    total += weight[batch]
    batch += 1
if best <= score:
    with open("_A2A_" + ID + ".py", "w") as fout:
        fout.write("# " + ID + " " + str(score) + "\n")
    subprocess.run("cmd /c echo # %COMPUTERNAME% %USERNAME%>>_A2A_" + ID + ".py")
    with open("_A2A_" + ID + ".py", "a") as fout:
        with open("_A2A_Solution.py") as code:
            fout.write(code.read())
print("Tentative score = " + str(score / total) + "/1", flush=True)
os._exit(0)
